### PR TITLE
fix(agw): Fixed SCTP abort issue by setting finite timeout in sctp_sendmsg

### PR DESCRIPTION
**Pull request by bhuvaneshne**
Signed-off-by: bhuvaneshne <bhuvaneshne@highway9networks.com>

## Summary

See bug description (https://github.com/magma/magma/issues/13115) for description, RCA and proposal of fix.

## Test Plan

Reproduce the issue using the bundled simulator (See: https://github.com/magma/magma/issues/13115)
Replace the sctpd executable
Restart sctp daemon (systemctl restart sctpd)
Wait for sctpd to get ready (See syslog)
Rerun the simulator and see that the issue is not reproducible again

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/sctpd/src/sctp_connection.cpp
Comment (ssanadhya) [inline lte/gateway/c/sctpd/src/sctp_connection.cpp:90]: Please add a comment stating that the 100 indicates a timetolive value of 100 ms.

**Review by ssanadhya — APPROVED** [review]
Lgtm, with minor comment. Please address before merging.

**Comment by ssanadhya** [comment]
@bhuvaneshne , please run [LTE integ tests](https://magma.github.io/magma/docs/next/lte/s1ap_tests) on this change and update the test plan.

**Comment by Neudrino** [comment]
This needs to be rebased on master, as there are foreign commits in here, which should not be on this PR.

**Review by Neudrino — DISMISSED** [review]
I am removing all the reviewers, as they are only on here due to bad commits in the branch!

**Comment by bhuvaneshne** [comment]
@Neudrino : Thanks for edit. I have moved this to draft. Will move it out once I fix.

**Comment by bhuvaneshne** [comment]
> This needs to be rebased on master, as there are foreign commits in here, which should not be on this PR.

Hi @Neudrino : I have fixed the commit. Please help review.

**Comment by Neudrino** [comment]
The correct approver groups for this PR would be @magma/approvers-agw-mme afaiks.
I added that approver groups and started the CI.

**Comment by ssanadhya** [comment]
> @bhuvaneshne , please run [LTE integ tests](https://magma.github.io/magma/docs/next/lte/s1ap_tests) on this change and update the test plan.

@bhuvaneshne , could you please confirm that LTE integ tests run fine on this PR? Please add that to the test plan list. 
Further, it looks like this one needs to be rebased on to latest master.

**Comment by bhuvaneshne** [comment]
> > @bhuvaneshne , please run [LTE integ tests](https://magma.github.io/magma/docs/next/lte/s1ap_tests) on this change and update the test plan.
> 
> @bhuvaneshne , could you please confirm that LTE integ tests run fine on this PR? Please add that to the test plan list. Further, it looks like this one needs to be rebased on to latest master.

@ssanadhya : I am yet to run the integ tests - Having trouble bringing up trfgen-server - It is throwing syntax error. Will update here once I get it running.

**Comment by bhuvaneshne** [comment]
> > > @bhuvaneshne , please run [LTE integ tests](https://magma.github.io/magma/docs/next/lte/s1ap_tests) on this change and update the test plan.
> > 
> > 
> > @bhuvaneshne , could you please confirm that LTE integ tests run fine on this PR? Please add that to the test plan list. Further, it looks like this one needs to be rebased on to latest master.
> 
> @ssanadhya : I am yet to run the integ tests - Having trouble bringing up trfgen-server - It is throwing syntax error. Will update here once I get it running.

Hi @ssanadhya , I had to revert file "traffic_messages.py" to older version for the trfgen-server to come up. However, the "make integ_test" is failing after about an hour into the run. This fails on the master branch too. Below are last few lines of the run
```
Waiting for 1 seconds for restart to complete                                                                                                         
Keys left in Redis (list should be empty)[                                                                                                            
                                                                                                                                                      
]                                                                                                                                                     
Entries in s1ap_imsi_map (should be zero): 0                                                                                                          
Entries left in hashtables (should be zero): 0                                                                                                        
Entries in mme_ueip_imsi_map (should be zero): 0                                                                                                      
E                                                                                                                                                     
======================================================================                                                                                
ERROR: Attach, detach, re-attach, send DL TCP data, and detach                                                                                        
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_attach_dl_tcp_data.py", line 73, in test_attach_detach_attach_dl_tcp_data
    test.verify()
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py", line 711, in verify
    raise RuntimeError(result.error)
RuntimeError: error - unable to connect to server: No route to host
-------------------- >> begin captured logging << --------------------
root: ERROR: Could not load magmad yml config for mconfig modules
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 16
root: INFO: Adding subscriber : IMSI001010000000001
root: DEBUG: s1ap response expected, received: 12, 12
root: DEBUG: s1ap message expected, received: 68, 68
root: DEBUG: s1ap response expected, received: 12, 12
root: DEBUG: s1ap message expected, received: 68, 68
root: INFO: Deleting subscriber : IMSI001010000000001
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 58.475s

FAILED (errors=1)
make: *** [Makefile:51: integ_test] Error 1
```

**Review by ajahl — APPROVED** [review]

**Comment by bhuvaneshne** [comment]
> > > > > @bhuvaneshne , please run [LTE integ tests](https://magma.github.io/magma/docs/next/lte/s1ap_tests) on this change and update the test plan.
> > > > 
> > > > 
> > > > @bhuvaneshne , could you please confirm that LTE integ tests run fine on this PR? Please add that to the test plan list. Further, it looks like this one needs to be rebased on to latest master.
> > > 
> > > 
> > > @ssanadhya : I am yet to run the integ tests - Having trouble bringing up trfgen-server - It is throwing syntax error. Will update here once I get it running.
> > 
> > 
> > Hi @ssanadhya , I had to revert file "traffic_messages.py" to older version for the trfgen-server to come up. However, the "make integ_test" is failing after about an hour into the run. This fails on the master branch too. Below are last few lines of the run
> > ```
> > Waiting for 1 seconds for restart to complete                                                                                                         
> > Keys left in Redis (list should be empty)[                                                                                                            
> >                                                                                                                                                       
> > ]                                                                                                                                                     
> > Entries in s1ap_imsi_map (should be zero): 0                                                                                                          
> > Entries left in hashtables (should be zero): 0                                                                                                        
> > Entries in mme_ueip_imsi_map (should be zero): 0                                                                                                      
> > E                                                                                                                                                     
> > ======================================================================                                                                                
> > ERROR: Attach, detach, re-attach, send DL TCP data, and detach                                                                                        
> > ----------------------------------------------------------------------
> > Traceback (most recent call last):
> >   File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_attach_detach_attach_dl_tcp_data.py", line 73, in test_attach_detach_attach_dl_tcp_data
> >     test.verify()
> >   File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py", line 711, in verify
> >     raise RuntimeError(result.error)
> > RuntimeError: error - unable to connect to server: No route to host
> > -------------------- >> begin captured logging << --------------------
> > root: ERROR: Could not load magmad yml config for mconfig modules
> > urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
> > urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 16
> > root: INFO: Adding subscriber : IMSI001010000000001
> > root: DEBUG: s1ap response expected, received: 12, 12
> > root: DEBUG: s1ap message expected, received: 68, 68
> > root: DEBUG: s1ap response expected, received: 12, 12
> > root: DEBUG: s1ap message expected, received: 68, 68
> > root: INFO: Deleting subscriber : IMSI001010000000001
> > --------------------- >> end captured logging << ---------------------
> > 
> > ----------------------------------------------------------------------
> > Ran 1 test in 58.475s
> > 
> > FAILED (errors=1)
> > make: *** [Makefile:51: integ_test] Error 1
> > ```
> 
> @bhuvaneshne , this could be a flake with the connection between the VMs. Can you try reloading all the VMs and re-running the tests? Further, the issue with "traffic_messages.py" should be fixed on master branch now, since we reverted the buggy PR.

Hi @ssanadhya : Yes, the issue with trfgen-server is resolved in latest master. The tests are failing midway (Ran about 2 hours this time) even after reprovisioning the VMs. Below is the last error report.
```
Entries in s1ap_imsi_map (should be zero): 0
Entries left in hashtables (should be zero): 0
Entries in mme_ueip_imsi_map (should be zero): 0
E
======================================================================
ERROR: Attach and send UL UDP data
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/test_attach_ul_udp_data.py", line 53, in test_attach_ul_udp_data
    test.verify()
  File "/home/vagrant/magma/lte/gateway/python/integ_tests/s1aptests/util/traffic_util.py", line 696, in verify
    raise RuntimeError(
RuntimeError: Cached results object is not a tuple : None
-------------------- >> begin captured logging << --------------------
root: ERROR: Could not load magmad yml config for mconfig modules
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 192.168.60.142:3333
urllib3.connectionpool: DEBUG: http://192.168.60.142:3333 "GET /stats/switches HTTP/1.1" 200 17
root: INFO: Adding subscriber : IMSI001010000000001
root: DEBUG: s1ap response expected, received: 12, 12
root: DEBUG: s1ap message expected, received: 68, 68
root: INFO: Deleting subscriber : IMSI001010000000001
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 1 test in 233.673s

FAILED (errors=1)
make: *** [Makefile:51: integ_test] Error 1

**Comment by ssanadhya** [comment]
@bhuvaneshne , this is the first test with traffic testing, so few things to check:
1. Is the traffic server running okay in your local setup?
2. Can you run this test individually, instead of as part of the sanity suite? You can do it with:

`make integ_test TESTS=s1aptests/test_attach_ul_udp_data.py`

@VinashakAnkitAman do we need to reload the VMs if one of the traffic tests fail?

2. Can you run this test on the master commit, which is the parent of your feature branch?

I am currently experiencing https://github.com/magma/magma/issues/13245 , hence cannot validate in my local setup.

**Comment by VinashakAnkitAman** [comment]
> @bhuvaneshne , this is the first test with traffic testing, so few things to check:
> 
> 1. Is the traffic server running okay in your local setup?
> 2. Can you run this test individually, instead of as part of the sanity suite? You can do it with:
> 
> `make integ_test TESTS=s1aptests/test_attach_ul_udp_data.py`
> 
> @VinashakAnkitAman do we need to reload the VMs if one of the traffic tests fail?
> 
> 2. Can you run this test on the master commit, which is the parent of your feature branch?
> 
> I am currently experiencing #13245 , hence cannot validate in my local setup.

Hi @ssanadhya All the cleanups happen in the scenario a test case fails including cleaning the interfaces and stopping the unclosed iperf instances. If any test case fails, including data test case, we need not reload the VMs. 

Hi @bhuvaneshne Although route should not be an issue here for uplink data test cases, can you please add route in the TRF Server VM with below command before running the test case?
`sudo ip route flush via 192.168.129.1 && sudo ip route add 192.168.128.0/24 via 192.168.129.1 dev eth2`

Now if you still have issues with local setup, since the branch has already been created and raised for PR, there is one more quick approach for you to verify the integ test directly via Github Actions:
- Go to your forked repo `https://github.com/<github_username>/magma/actions/workflows/lte-integ-test.yml`
- Click `Run workflow` button on the right side and select the correct branch, then click the `Run workflow button`.
- Then follow the latest execution report downside

**Comment by VinashakAnkitAman** [comment]
@bhuvaneshne By any chance you got the integ test working over this PR?

**Comment by bhuvaneshne** [comment]
> > @bhuvaneshne By any chance you got the integ test working over this PR?
> 
> @bhuvaneshne By any chance you got the integ test working over this PR?
> 
> > > > @bhuvaneshne , this is the first test with traffic testing, so few things to check:
> > > > 
> > > > 1. Is the traffic server running okay in your local setup?
> > > > 2. Can you run this test individually, instead of as part of the sanity suite? You can do it with:
> > > > 
> > > > `make integ_test TESTS=s1aptests/test_attach_ul_udp_data.py`
> > > > @VinashakAnkitAman do we need to reload the VMs if one of the traffic tests fail?
> > > > 
> > > > 2. Can you run this test on the master commit, which is the parent of your feature branch?
> > > > 
> > > > I am currently experiencing #13245 , hence cannot validate in my local setup.
> > > 
> > > 
> > > Hi @ssanadhya All the cleanups happen in the scenario a test case fails including cleaning the interfaces and stopping the unclosed iperf instances. If any test case fails, including data test case, we need not reload the VMs.
> > > Hi @bhuvaneshne Although route should not be an issue here for uplink data test cases, can you please add route in the TRF Server VM with below command before running the test case? `sudo ip route flush via 192.168.129.1 && sudo ip route add 192.168.128.0/24 via 192.168.129.1 dev eth2`
> > > Now if you still have issues with local setup, since the branch has already been created and raised for PR, there is one more quick approach for you to verify the integ test directly via Github Actions:
> > > ```
> > > * Go to your forked repo `https://github.com/<github_username>/magma/actions/workflows/lte-integ-test.yml`
> > > 
> > > * Click `Run workflow` button on the right side and select the correct branch, then click the `Run workflow` button.
> > > 
> > > * Then follow the latest execution report downside
> > > ```
> > 
> > 
> > Thanks @VinashakAnkitAman , I rebased and reran the test on a new environment. It failed again but in a different spot. Let me try the "run workflow"
> > ```
> > Deleting all the uplink/downlink flows
> > Keys left in Redis (list should be empty)[
> >  IMSI001010000000001:SPGW
> > IMSI001010000000001.magma.ipv4,ipv4:mobilityd_ipdesc_record
> > IMSI001010000000001:MME
> > 
> > nb_ue_attached: 1 
> > ]
> > Entries in s1ap_imsi_map (should be zero): 1
> > Entries left in hashtables (should be zero): 3
> > Entries in mme_ueip_imsi_map (should be zero): 1
> > The test has failed. Restarting Sctpd for cleanup
> > Waiting for 30 seconds for restart to complete
> > Waiting for 29 seconds for restart to complete
> > Waiting for 28 seconds for restart to complete
> > Waiting for 27 seconds for restart to complete
> > Waiting for 26 seconds for restart to complete
> > Waiting for 25 seconds for restart to complete
> > Waiting for 24 seconds for restart to complete
> > Waiting for 23 seconds for restart to complete
> > Waiting for 22 seconds for restart to complete
> > Waiting for 21 seconds for restart to complete
> > Waiting for 20 seconds for restart to complete
> > Waiting for 19 seconds for restart to complete
> > Waiting for 18 seconds for restart to complete
> > Waiting for 17 seconds for restart to complete
> > Waiting for 16 seconds for restart to complete
> > Waiting for 15 seconds for restart to complete
> > Waiting for 14 seconds for restart to complete
> > Waiting for 13 seconds for restart to complete
> > Waiting for 12 seconds for restart to complete
> > Waiting for 11 seconds for restart to complete
> > Waiting for 10 seconds for restart to complete
> > Waiting for 9 seconds for restart to complete
> > Waiting for 8 seconds for restart to complete
> > Waiting for 7 seconds for restart to complete
> > Waiting for 6 seconds for restart to complete
> > Waiting for 5 seconds for restart to complete
> > Waiting for 4 seconds for restart to complete
> > Waiting for 3 seconds for restart to complete
> > Waiting for 2 seconds for restart to complete
> > Waiting for 1 seconds for restart to complete
> > Keys left in Redis (list should be empty)[
> >   
> > ]
> > Entries in s1ap_imsi_map (should be zero): 0
> > Entries left in hashtables (should be zero): 0
> > Entries in mme_ueip_imsi_map (should be zero): 0
> > -------------- generated xml file: /var/tmp/test_results/test_tau_periodic_active.xml ---------------
> > ====================================== short test summary info ======================================
> > FAILED s1aptests/test_tau_periodic_active.py::TestTauPeriodicActive::test_tau_periodic_active - As...
> > !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
> > =================================== 1 failed in 117.67s (0:01:57) ===================================
> > make: *** [Makefile:51: integ_test] Error 1
> > ```
> 
> @bhuvaneshne By any chance you got the integ test working over this PR?

Hi @VinashakAnkitAman : the test reported failure. I did not retry once again - I will give it a shot once again and update.

**Comment by ssanadhya** [comment]
@bhuvaneshne , it will be great to have this fix in the release 1.8 (which was branched out last week). If you could please confirm that LTE integration tests are passing on this PR, we can merge and backport this change to the official release.

@VinashakAnkitAman , can you also test the changes of this PR in your local setup?

**Comment by VinashakAnkitAman** [comment]
> @bhuvaneshne , it will be great to have this fix in the release 1.8 (which was branched out last week). If you could please confirm that LTE integration tests are passing on this PR, we can merge and backport this change to the official release.
> 
> @VinashakAnkitAman , can you also test the changes of this PR in your local setup?

Sure @ssanadhya I am on leave on Tuesday for some sudden personal work.. I will re-verify in my local setup as well on Wednesday morning..

**Comment by bhuvaneshne** [comment]
> > @bhuvaneshne , it will be great to have this fix in the release 1.8 (which was branched out last week). If you could please confirm that LTE integration tests are passing on this PR, we can merge and backport this change to the official release.
> > @VinashakAnkitAman , can you also test the changes of this PR in your local setup?
> 
> Sure @ssanadhya I am on leave on Tuesday for some sudden personal work.. I will re-verify in my local setup as well on Wednesday morning..

Hi @ssanadhya , Below are the test cases that fail even after retries:
image

I can rebase and give it one more try

**Comment by bhuvaneshne** [comment]
@ssanadhya , Finally got a clean run (Had to retry couple of test cases). Attaching the test results
[test_results.zip](https://github.com/magma/magma/files/9355737/test_results.zip)
.

**Comment by ssanadhya** [comment]
Thanks for confirming @bhuvaneshne ! I will merge this once CI checks are complete. 

> @ssanadhya , Finally got a clean run (Had to retry couple of test cases). Attaching the test results [test_results.zip](https://github.com/magma/magma/files/9355737/test_results.zip) .

**Comment by nstng** [comment]
already in 1.8 - setting respective label

**Pull request merged**
